lib/hanami/application/container/boot/logger.rb

```diff
@@ -2,7 +2,6 @@
 
 Hanami.application.register_bootable :logger do
   start do
-    require "hanami/logger"
-    register :logger, Hanami::Logger.new(**Hanami.application.configuration.logger)
+    register :logger, Hanami.application.configuration.logger.instance
```

Comment:
@timriley It seems a SRP violation to have a config object (logger settings) to also build the target (the logger itself). Can't we keep this LOC as it was?

Reply:
Yeah, on reflection I think you're right. Thanks for raising this.

I agonised over this one for a while. There's a couple of in-tension criteria I wanted to satisfy with this `configuration.logger` object:

- Allow particular aspects of the standard logger to be configured without the user having to pre-initialize a whole replacement object (e.g. the user who just wants to change the log level, perhaps also adjusting it in different Hanami environments, but leave everything else the same)
- But also allow the user to supply their own pre-initialized logger object of their choosing, and have that override those finer-grained `logger_class`/`options` settings

This is what led me to the current implementation of `Configuration#logger=` (which sets the pre-initialized logger object) as well as the `instance` attribute on which that pre-initialized logger is stored. Then from there, given the job of _constructing_ the logger is now non-trivial and contingent on several different settings, which should be checked in a particular order, I thought it would be good to encapsulate that behaviour alongside those settings, so they can all evolve together when needed.

If we removed this, it's true that it would make `Configuration::Logger` "purer" and more single-responsibility, but it does mean that this logger initialisation logic needs to go elsewhere and stay in sync.

I suppose if it _only_ ends up here in this bootable component file, and doesn't have to be repeated in other places, then I think we should be fine. I'll change it back now.

While we're discussing this, do you have any other ideas about about how we could satisfy those two different criteria I listed above, or do you think the way I've done it is reasonable enough? Is the second of those two criteria even worth supporting? (I think so, if possible, since I'd like to make Hanami as flexible as possible for those you need to go off the beaten track, so long as it doesn't compromise the experience for the other 99% of users)

Reply:
Another option here could be to move the logger options back into the top-level configuration class, but from my perspective, as soon as there are >2 settings at the top level with the same underscored "namespace", e.g. `logger_this`, `logger_that`, and `logger_another`, it's a clear sign that we could make things clearer to understand and maintain by grouping them together under a nested configuration object.

Reply:
I've just pushed [`9226d1a`](https://github.com/hanami/hanami/pull/1107/commits/9226d1a20596b70c74292c58f29d359a88badd1d), which removes the building of the instance from the `Configuration::Logger` class.

Reply:
On the topic of allowing a user to supply their own pre-initialized logger, I did just write this comment in the other thread here (https://github.com/hanami/hanami/pull/1107#discussion_r634323665), which describes how you can override the whole `logger` bootable component to register your own logger instance in the Hanami application container. This is already a viable path for a user who wants to do this, so perhaps we _don't_ need to account for it in this logger configuration object? That said, that's a pretty advanced technique, and we already e.g. support providing an inflector object via configuration, so perhaps allowing it for the logger does make sense too?

Reply:
@timriley Can't the responsibility of instantiating the logger be delegated to `Hanami::Configuration`?

Pseudo code to explain myself.

In the bootable component:

```ruby
register :logger, Hanami.application.configuration.application_logger
```

In configuration:

```ruby
setting :logger, Configuration::Logger.new, cloneable: true

def logger=(logger_instance)
  config.logger = Configuration::Logger.new(logger_instance)
end

# @api private
def application_logger
  manually_assigned_logger || Hanami::Logger.new(*logger_settings)
end
```

Reply:
@jodosha Thanks for the great ideas :) I've just incorporated this in [`25db2b9` (#1107)](https://github.com/hanami/hanami/pull/1107/commits/25db2b94b2e67246042d794d7a2216c597375d1e).

If you take a look at that commit, there are a couple of subtle differences I wanted to note:

1. I've called the method `#logger_instance` instead of `#application_logger`. In this case I feel it better describes what it actually returns. The application prefix felt a little redundant, and the _true_ "application logger" is the one that we register in `lib/hanami/application/container/boot/logger.rb`.
2. I also changed it so it doesn't memoize a logger built from the `logger` settings. The reason for this is that we want to make sure settings changes made to `config.logger.*` are _always_ respected when it comes time to get the full logger object. If we memoize, then as soon as we access the `logger_instance`, any subsequent changes to `configuration.logger.*` will just be ignored, which would be unexpected and potentially lead to bugs or unwanted behaviour.

Does that look OK to you?

All I think I have left here is just to make sure there's a test to cover this adjusted behaviour, then I think this'll be ready to merge.